### PR TITLE
#368 equation node on close behavior, #379 updating equation editor b…

### PR DESCRIPTION
**Pull request by riteshsamala**
…ased on model, bug: updating new equation editor on add equation

PR Todos: 
- [x] Make it so the node editor done button doesn't close the node if the equation has a syntax error--instead display the message in the message box and leave the editor open so the author can fix it.

### CODE REVIEW

**Comment by jonwetzel** [comment]
Critical bug found.

1. Click add equation
2. Select a schema
3. Click "Done"  At first the node does not close.  Incorrect equation syntax appears in the box.  (Not sure if this is expected, will talk with Kurt.)  
4. Click "Done" again
5. Click "Add Equation" and check console for error.
After this "Add Equation" no longer works and clicking on the nodes does not open the node editor due to console error:

```
Uncaught Error: parse error [column 1]: parity
    at Parser.error_parsing (parser.js?1536168231565:778)
    at Parser.parse (parser.js?1536168231565:765)
    at Function.Parser.parse (parser.js?1536168231565:555)
    at Object.<anonymous> (equation.js?1536168231565:571)
    at Object.forEach (array.js?1536168231565:247)
    at Object.getVariableStrings (equation.js?1536168231565:570)
    at Object.fillVariableNames (con-author.js?1536168231565:1408)
    at Object.initialControlSettings (con-author.js?1536168231565:928)
    at Object.showNodeEditor (controller.js?1536168231565:558)
    at Object.advice (aspect.js?1536168231565:114)```

**Comment by jonwetzel** [comment]
@riteshsamala Did you make any changes to the math solver for this?  I pulled the latest but it's still not working.

**Comment by jonwetzel** [comment]
@riteshsamala Found another critical bug.  I filled out an equation properly but it gives "Incorrect equation syntax".

eq-editor-bug2

**Comment by jonwetzel** [comment]
@riteshsamala As discussed, make the following fix: 
- [ ] Make it so the node editor done button doesn't close the node if the equation has a syntax error--instead display the message in the message box and leave the editor open so the author can fix it.

File: www/js/con-author.js [hunk 1]
```diff
@@ -588,6 +588,14 @@ define([
 			var logObj = {};
 			if(model && model == "authored"){
 				var directives = [];
+				//before parsing the equation we need to check if all variables are present and none of the variable fields is left empty
```
Comment (jonwetzel) [inline www/js/con-author.js]: Looks good, but I found a bug related to this function--if the user clicks "Delete Node" this check still triggers.  The equation done handler shouldn't need to be called on delete I think...

File: www/js/equation.js [hunk 2]
```diff
@@ -556,6 +556,28 @@ define([
...
+			if(eqs.length != 2){
+				throw new Error("Wrong number of equal to symbols in the equation");
+			}
+			var expressions = [];
+			try{
+				array.forEach(eqs, function(eq, count){
+					expressions[count] = Parser.parse(eq);
+				}, this);
+			}catch(e){
+				throw e;
```
Comment (jonwetzel) [inline www/js/equation.js]: Why a try-catch block if we're only just going to throw the error anyway?

Comment (jonwetzel) [reply www/js/equation.js]: Logging the error sounds good to me.

Comment (riteshsamala) [reply www/js/equation.js]: Can we push all the logging to a separate task later and leave a note there now that we have more prioritized tasks ?

Comment (jonwetzel) [reply www/js/equation.js]: Sure.

**Comment by jonwetzel** [comment]
New bug found:

* Add a DRT where D1 = R1 * T1
* Click Done
* Click on DRT eqn to edit it
* Change D1 to Q1
* Click Done

The equation is valid, but the node editor does not close on the first click.

**Comment by jonwetzel** [comment]
New bug found:  Deleting an equation doesn't work

* Add a DRT where D1 = R1 * T1
* Click Done
* Click on DRT eqn to edit it
* Click Delete Node

Node editor closes but the node is not deleted.

**Comment by jonwetzel** [comment]
We tried but neither of the new bugs were reproducible on other machines.  (This happened on the one in BYAC 110)

**Pull request merged**